README.md
# DesyncOscillators
Here I am trying to model the behaviour of simple amplitude-phase oscillators.
The main model for a single oscillator is modified Poincare oscillator.
I am trying to simulate the effects of amplitude relaxation rate, twist (parameter linking amplitude and period), 
coupling between oscillators (modeled as mean-field coupling), 


history
# -*- coding: utf-8 -*-
# *** Spyder Python Console History Log ***

## ---(Thu Feb 22 11:39:07 2018)---
sin(2)
2+3
alph = 'abcdefg'
alph[:2]
alph[2:]
10//2
10/2
100//24
100/24
import matplotlib.pyplot*
import matplotlib.pyplot
import matplotlib.pyplot as mpl
runfile('/home/kalashnikov/.config/spyder-py3/temp.py', wdir='/home/kalashnikov/.config/spyder-py3')
fib(2000)
range(0,10)
list[0:10:2]
range(0:10:2)
range(0,10,2)
range(5, 10)
list(range(0,10,2))
mpl.plot(fib(2000),range(0,len(fib(2000))))
len(fib(2000))
mpl.plot(fib(2000),range(0,len(fib(2000)+1)))
mpl.plot(fib(2000),range(0,len(fib(2000)+'1')))
mpl.plot(fib(2000),range(0,len(fib(2000)+[1])))
mpl.plot(fib(2000),)
mpl.plot(fib(2000))
mpl.plot(fib(10000))
mpl.plot(fib(1000000))
mpl.plot(fib(1000000000))
import numpy as np
np.arange?
np.lispace?
np.linspace?
n = np.arange(0,1,0.1)
n
n[0]
n[3]
3[5]
n[5]
n[2]
n[7]
np.diff?
n
n1=np.diff(n)
n1
1/n1
runfile('/home/kalashnikov/.config/spyder-py3/temp.py', wdir='/home/kalashnikov/.config/spyder-py3')
tonicNeuron(1000,1)
tonicNeuron(10,1)
tonicNeuron(10,2)
tonicNeuron(100,1)
tonicNeuron(100,2)
tonicNeuron(100,3)
tonicNeuron(100,10)
tonicNeuron(100,0.1)
tonicNeuron(100,0.5)
tonicNeuron(100,0.3)
tonicNeuron(100,0.4)
tonicNeuron(100,0.1)
tonicNeuron(100,0.01)
tonicNeuron(100,10)
print(tonicNeuron(100,10))
runfile('/home/kalashnikov/.config/spyder-py3/temp.py', wdir='/home/kalashnikov/.config/spyder-py3')
tonicNeuron(100,10)
tonicNeuron(100,1)
tonicNeuron(10,0.1)
runfile('/home/kalashnikov/.config/spyder-py3/temp.py', wdir='/home/kalashnikov/.config/spyder-py3')
tonicNeuron(10,0.1)
runfile('/home/kalashnikov/.config/spyder-py3/temp.py', wdir='/home/kalashnikov/.config/spyder-py3')
a = tonicNeuron(10,0.1)
a[10]
a[11]
a[200]
a[20]
a = tonicNeuron(10,0.01)
a
a[0]
a[1]
a[12]
a[10]
round?
round(10,3452,2)
round(10.3452,2)
runfile('/home/kalashnikov/.config/spyder-py3/temp.py', wdir='/home/kalashnikov/.config/spyder-py3')
a = tonicNeuron(10,0.1)
runfile('/home/kalashnikov/.config/spyder-py3/temp.py', wdir='/home/kalashnikov/.config/spyder-py3')
a = tonicNeuron(10,0.1)
a = tonicNeuron(10,0.01)
a = tonicNeuron(10,0.001)
a = tonicNeuron(10,0.0001)
a = tonicNeuron(2,0.00001)
import numpy.random as npr
npr.random(1)
npr.random(10)
npr.random?
npr.random(3)
npr.random(3)*5
npr.random(3)*10-5
npr.random(100)*10-5
runfile('/home/kalashnikov/.config/spyder-py3/temp.py', wdir='/home/kalashnikov/.config/spyder-py3')
UniformNeuron(3)
runfile('/home/kalashnikov/.config/spyder-py3/temp.py', wdir='/home/kalashnikov/.config/spyder-py3')
UniformNeuron(3)
runfile('/home/kalashnikov/.config/spyder-py3/temp.py', wdir='/home/kalashnikov/.config/spyder-py3')
UniformNeuron(100)
runfile('/home/kalashnikov/.config/spyder-py3/temp.py', wdir='/home/kalashnikov/.config/spyder-py3')
UniformNeuron(100)
UniformNeuron(1000)
UniformNeuron(100,2)
UniformNeuron(100,3)
UniformNeuron(100,2)
UniformNeuron(100,4)
UniformNeuron(100,1)
UniformNeuron(100,0.1)
UniformNeuron(100,1)
UniformNeuron(100,10)
UniformNeuron(100,10,0.1)
UniformNeuron(100,10,0.2)
UniformNeuron(100,1,0.2)
UniformNeuron(100,0.1,0.2)
runfile('/home/kalashnikov/.config/spyder-py3/temp.py', wdir='/home/kalashnikov/.config/spyder-py3')
UniformNeuron(100,0.1,0.2)
UniformNeuron(1000,0.1,0.2)
UniformNeuron(1000,0.1)
UniformNeuron(1000,0.2)
UniformNeuron(1000,0.3)
UniformNeuron(1000,10)
UniformNeuron(1000,0.01)
UniformNeuron(1000,1,0.01)
UniformNeuron(1000,1,0.02)
UniformNeuron(1000,1,0.03)
UniformNeuron(1000,1,0.1)
UniformNeuron(1000,1,0.01)
UniformNeuron(1000,1,0.2)
UniformNeuron(1000,1,0.3)
UniformNeuron(1000,1,0.5)
UniformNeuron(1000,1,1)
UniformNeuron(1000,1,2)
UniformNeuron(1000,1,3)
UniformNeuron(1000,1,10)
UniformNeuron(10000,1,10)
UniformNeuron(10000,1,`)
UniformNeuron(10000,1,1)
UniformNeuron(10000,1,0.1)
UniformNeuron(10000,1,0.01)
tonicNeuron(100)
tonicNeuron(100,1)
tonicNeuron(100,2)
tonicNeuron(100,10)
runfile('/home/kalashnikov/.config/spyder-py3/temp.py', wdir='/home/kalashnikov/.config/spyder-py3')
tonicNeuron(100,10)
tonicNeuron(100,1)
tonicNeuron(10,1)
tonicNeuron(10,2)
tonicNeuron(10,0.1)
tonicNeuron(10,0.2)
tonicNeuron(10,0.3)
npr.poisson(1)
npr.poisson(2)
npr.poisson?
npr.poisson(0)
npr.poisson(1)
npr.poisson(2)
npr.poisson(2, 100)
plt.plot(npr.poisson(1,100))
plt.plot(npr.poisson(2,100))
plt.plot(npr.poisson(5,100))
plt.plot(npr.poisson(100,100))
plt.plot(npr.poisson(1000,100))
plt.plot(npr.poisson(1000,1000))
plt.plot(npr.poisson(100,1000))
plt.plot(npr.poisson(10,1000))
plt.plot(npr.poisson(1,1000))
plt.plot(npr.poisson(2,1000))
plt.plot(npr.poisson(1000,1000))
plt.plot(npr.poisson(10000,1000))
plt.plot(npr.poisson(10000,10000))
plt.plot(npr.poisson(10000,10000), '.')
plt.plot(npr.poisson(1000,1000), '.')
npr.poisson(2, 100)
plt.hist(npr.poisson(2,100)))
plt.hist(npr.poisson(2,100)),1)
a=npr.poisson(2, 100)
plt.hist(a)
a=npr.poisson(2, 1000)
a
plt.hist(a)
a=npr.poisson(100, 1000)
a
plt.hist(a)
a=npr.poisson(1, 1000)
plt.hist(a)
a=npr.poisson(100, 1000)
plt.hist(a,bin=100)
plt.hist(a,bins=100)
plt.hist(a,bins=30)
plt.hist(a,bins=35)
plt.hist(a,bins=40)
plt.hist(a,bins=100)
plt.hist(a,bins=1000)
plt.hist(a,bins=200)
a=npr.poisson(1, 10000)
a=npr.poisson(2, 10000)
plt.hist(a,bins=200)
plt.hist(a,bins=100)

## ---(Thu Mar 15 17:35:35 2018)---
pylab?
scipy?
import scipy as sp
runfile('/home/kalashnikov/Code/My first Poincare oscillator.py', wdir='/home/kalashnikov/Code')
help ("modules")

## ---(Thu Mar 15 17:56:24 2018)---
scipy?
runfile('/home/kalashnikov/Code/My first Poincare oscillator.py', wdir='/home/kalashnikov/Code')
help ("modules")
import scipy.integrate as spi
runfile('/home/kalashnikov/Code/My first Poincare oscillator.py', wdir='/home/kalashnikov/Code')
clear all
runfile('/home/kalashnikov/Code/My first Poincare oscillator.py', wdir='/home/kalashnikov/Code')
spi.odeint?
runfile('/home/kalashnikov/Code/My first Poincare oscillator.py', wdir='/home/kalashnikov/Code')
scipy?

PoincareOsc.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Mar 15 17:37:23 2018

@author: kalashnikov
"""
from scipy.integrate import odeint
import numpy as np
import matplotlib.pyplot as plt

# function that returns dy/dt
def model2 (vector,t,alpha, A, omega, twist):
    x = vector[0]
    y = vector[1]
    dxdt = x*alpha*(A-np.sqrt(x**2 + y**2)) - y*(omega + twist*(A - np.sqrt(x**2 + y**2)))
    dydt = y*alpha*(A-np.sqrt(x**2 + y**2)) + x*(omega + twist*(A - np.sqrt(x**2 + y**2)))
    dzdt = [dxdt, dydt]
    return dzdt

# time points
stepsize = 1
hours = 30
start = 0        
t = np.linspace(start,hours,hours*stepsize) # generates the timepoints from start, hours-long with a stepsize
#notice, that np.linspace(start, hours, stepsize) doesn't really provide you with a stepsize you want it to be


# initial conditions
state0 = [1,1]
state01 = [0,1]
state02 = [4,4]
state03 = [1.5,1]

# sets of parameters
params = (0.1,1,(np.pi*2)/24, 0.5) #relaxation rate (alpha/lambda); amplitude (A); angle speed (omega); twist
params1 = (1,1,(np.pi*2)/24, 0)
params2 = (0.5,1,(np.pi*2)/24, 0.5)
params3 = (1,1,(np.pi*2)/24, 0.5)
params4 = (0.1,1,(np.pi*2)/24, 0.5)

# solves ODEs
#here are ODEs with different parameters but same initial conditions
x0 = odeint(model2, state02, t, args=(params))
#x1 = odeint(model2, state02, t, args=(params1))
#x2 = odeint(model2, state02, t, args=(params2))
#x3 = odeint(model2, state02, t, args=(params3))
#x4 = odeint(model2, state02, t, args=(params4))



# Here we extract the 0-crossings (only changes from + to -, without from - to +), mins and maxes

# Should I use y or x when finding the 0-crossings, mins and maxs? 
# Does it even matter? I wouldn't think it does
#the indices of x-values just before the crossings
zeroCrossInd = np.where(np.diff(np.sign(x0[:,0])))[0]
# the values themselves, should be really close to 0
zeroCrossVal = (x0[zeroCrossInd,0]+x0[zeroCrossInd+1,0])/2
#the approximate times of actual crossings
zeroCrossT = (t[zeroCrossInd]+t[zeroCrossInd+1])/2 # takes the time in between two x-values of opposing signs
period = np.diff(zeroCrossT)





# Local extrema
extrVal = [] # values
extrT = [] # timepoints for respective values

# Looking for local maxima, minima
# Maybe np.diff(np.sign(np.diff(x0[:,0])))? When result of np.diff() changes the sign - it's when the max of min occured 
diff = np.diff(np.sign(np.diff(x0[:,0])))
for i in range(len(diff)):
    if diff[i]!=0:
        extrVal.append(np.mean(x0[i:i+2,0]))
        extrT.append(np.mean(t[i:i+2]))





# different lists for minima and maxima        
mins=[]
maxs=[]
for i in extrVal:
    if i<0:
        mins.append(i)
    else:
        maxs.append(i)

fold = maxs[-1]/mins[-1]



# _____________________________________________
# _____________________________________________
# PLOTTING
# time-series
plt.figure(figsize=(12,5))
#plt.plot(t, x0[:,0], 'o', label = 'x from x0')
#plt.plot(t, x4[:,0], label = 'x from x4')
#plt.plot(zeroCrossT, np.zeros(len(zeroCrossT)), 'r+', label = 'zero crossings of x-coordinate of x0') 
#plt.plot(extrT,extrVal,'-v')
#plt.xlim(start-1, hours-start+1) #shows the x from a little bit before the start of timepoints, to a little bit after the end
#plt.ylim(-2.5,4.5)

plt.legend()
plt.show()







"""
So the next step is to identify the functions that need to be "functionized", in order o separate the methods from the data itself
"""







#______________________________________________

# TO-DO for tomorrow, 27th March
# 1) Extract the 0-crossings, maxs and mins as 3 different lists to compute then magnitude.
# 2) Find some nice way to make a lot of graphs without changing everything in all of them
#  2.a) Labels should automatically reflect the changes in the variables.
#  2.b) Find how you can make a plot bigger, jusr bigger in all dimensions, witout manually assigning plt.figure(figsize=(x,y))
#  2.c) plt.ylim and plt.xlim should automatically be a little bit larger than the max and min values.
#  2.d) Phase portraits should be plotted simultaneously with time-series and magnitude/period/phase, 
#       maybe on the same graphs, maybe on the different, that needs to be thought through.
#  2.e) Get rid of "Reloaded modules".






"""
# Another implementation of zeroCross
zeroCross=[]
for i in range(len(x0[:,0]-1)):
    if (x0[i,0]>0 and x0[i+1,0]<=0):
        zeroCross.append((t[i]+t[i+1])/2) # we append to zeroCross the approx. time of 0-crossing

#here are same ODEs with different initial conditions
x0 = odeint(model2, state0, t, args=(params))
x1 = odeint(model2, state01, t, args=(params))
x2 = odeint(model2, state02, t, args=(params))
x3 = odeint(model2, state03, t, args=(params))


# Phase-plots
plt.figure(figsize=(10,10))
plt.plot(x0[:,0],x0[:,1],'r-', linewidth=2, label = 'relax.rate = 0.5, twist=0')
plt.plot(x1[:,0],x1[:,1],'g-', linewidth=2, label = 'relax.rate = 1, twist=0')
plt.plot(x2[:,0],x2[:,1],'b-', linewidth=2, label = 'relax.rate = 0.5, twist=0.5')
plt.plot(x3[:,0],x3[:,1],'k-', linewidth=2, label = 'relax.rate = 1, twist=0.5')
#plt.plot(x4[:,0],x4[:,1],'m-', linewidth=2, label = 'relax.rate = 0.1, twist=0.5')

plt.xlim(-1.5,2.5)
plt.ylim(-1.5,2.5)
plt.xlabel('x(t)')
plt.ylabel('y(t)')
"""










